src/v/kafka/client/topic_cache.cc

```diff
@@ -38,9 +58,15 @@ void topic_cache::apply(
               partition_data{
                 .leader = p.leader_id, .leader_epoch = p.leader_epoch});
         }
+        if (t.topic_id != model::topic_id{}) {
+            // Topic ID must be present in Metadata response v12+
+            cache_t.topic_id = t.topic_id;
+            new_topic_ids[t.topic_id] = t_name;
+        }
     }
 
     std::exchange(_topics, std::move(new_cache));
+    std::exchange(_topic_ids, std::move(new_topic_ids));
```

Comment:
During a rolling upgrade, it's possible to connect to a newer broker with Topic IDs, and then a broker that has an older metadata version, which may not include topic IDs. It might be useful to have a slightly more complex algorithm here, that combines, rather than completely swaps out.